gtdynamics/dynamics/Chain.cpp

```diff
@@ -85,4 +85,80 @@ Pose3 Chain::poe(const Vector &q, boost::optional<Pose3 &> fTe,
   }
   return poe;
 }
+
+gtsam::Vector3 Chain::DynamicalEquality3(const gtsam::Vector6 &wrench,
+                                         const gtsam::Vector3 &angles,
+                                         const gtsam::Vector3 &torques,
+                                         gtsam::OptionalJacobian<3, 6> J0,
+                                         gtsam::OptionalJacobian<3, 3> J1,
+                                         gtsam::OptionalJacobian<3, 3> J2) {
+  Matrix J;
+  poe(angles, boost::none, J);
+  if (J0) {
+    // derivative of difference with respect to wrench
+    *J0 = J.transpose();
+  }
+  if (J1) {
+    // derivative of difference with respect to angles
+    // NOT COMPLETE
+    *J1 = gtsam::I_3x3;
+  }
+  if (J2) {
+    // derivative of difference with respect to torques
+    *J2 = -gtsam::I_3x3;
+  }
+
+  return (J.transpose() * wrench - torques);
+}
+
+// Helper function to create expression with a vector, used in
+// ChainConstraint3.
+gtsam::Vector3 MakeVector3(const double &value0, const double &value1,
+                           const double &value2,
+                           gtsam::OptionalJacobian<3, 1> J0 = boost::none,
+                           gtsam::OptionalJacobian<3, 1> J1 = boost::none,
+                           gtsam::OptionalJacobian<3, 1> J2 = boost::none) {
+  gtsam::Vector3 q;
+  q << value0, value1, value2;
+  if (J0) {
+    *J0 << 1.0, 0.0, 0.0;
+  }
+  if (J1) {
+    *J1 << 0.0, 1.0, 0.0;
+  }
+  if (J2) {
+    *J2 << 0.0, 0.0, 1.0;
+  }
+  return q;
+}
+
+gtsam::Vector3_ Chain::ChainConstraint3(
+    const std::vector<JointSharedPtr> &joints, const gtsam::Key wrench_key,
+    size_t k) {
+  // Get Expression for wrench
+  gtsam::Vector6_ wrench(wrench_key);
+
+  // Get expression for joint angles as a column vector of size 3.
+  gtsam::Vector3_ angles(MakeVector3,
+                         gtsam::Double_(JointAngleKey(joints[0]->id(), k)),
+                         gtsam::Double_(JointAngleKey(joints[1]->id(), k)),
+                         gtsam::Double_(JointAngleKey(joints[2]->id(), k)));
+
+  // Get expression for joint torques as a column vector of size 3.
+  gtsam::Vector3_ torques(MakeVector3,
+                          gtsam::Double_(TorqueKey(joints[0]->id(), k)),
```

Comment:
copy/pasta: just add
```
gtsam Double_ angle0(JointAngleKey(joints[0]->id(), k)), angle1(JointAngleKey(joints[1]->id(), k)), angle2(JointAngleKey(joints[2]->id(), k));
```

above?

Reply:
PS Why not just have a Vector3 in the factor graph?

Reply:
done

As we talked, I need the vector to hold 3 different keys (for now).
later I will incorporate a new type of key for the chain.